File: src/main/java/cephetir/skyskipped/commands/SkySkippedCommand.java
/*
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * Version 2, December 2004
 *
 * Copyright (C) 2021 Cephetir
 *
 * Everyone is permitted to copy and distribute verbatim or modified
 * copies of this license document, and changing it is allowed as long
 * as the name is changed.
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 */

package cephetir.skyskipped.commands;

import cephetir.skyskipped.SkySkipped;
import cephetir.skyskipped.config.Cache;
import gg.essential.universal.UChat;
import gg.essential.universal.UScreen;
import gg.essential.universal.wrappers.message.UTextComponent;
import net.minecraft.command.CommandBase;
import net.minecraft.command.ICommandSender;
import net.minecraft.event.ClickEvent;
import net.minecraft.event.HoverEvent;
import net.minecraft.util.BlockPos;

import java.util.Collections;
import java.util.List;

public class SkySkippedCommand extends CommandBase {

    @Override
    public String getCommandName() {
        return "skyskipped";
    }

    @Override
    public List<String> getCommandAliases() {
        return Collections.singletonList("sm");
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "/" + getCommandName() + " help";
    }

    @Override
    public int getRequiredPermissionLevel() {
        return 0;
    }

    @Override
    public List<String> addTabCompletionOptions(ICommandSender sender, String[] args, BlockPos pos) {
        if (args.length == 1) {
            return getListOfStringsMatchingLastWord(args, "github", "crit", "help", "gui");
        }
        return null;
    }

    @Override
    public void processCommand(ICommandSender sender, String[] args) {
        if (args.length == 0) {
            UScreen.displayScreen(SkySkipped.config.gui());
        } else if (args[0].equals("gui")) {
            UScreen.displayScreen(SkySkipped.config.gui());
        } else if (args[0].equals("github")) {
            UTextComponent text = new UTextComponent("§cSkySkipped §f:: §eGithub: §fhttps://github.com/Cephetir/SkySkipped");
            text.setHover(HoverEvent.Action.SHOW_TEXT, "§9Open URL in browser.");
            text.setClick(ClickEvent.Action.OPEN_URL, "https://github.com/Cephetir/SkySkipped");
            sender.addChatMessage(text);
        } else if (args[0].equals("crit")) {
            UTextComponent text = new UTextComponent("§cSkySkipped §f:: §eLast Crit: §f" + Cache.lastCrit);
            sender.addChatMessage(text);
        } else if (args[0].equals("help")) {
            UChat.chat("§cSkySkipped §f:: §eUsage:\n" +
                    "§cSkySkipped §f:: §e/sm §for§e /sm gui §f- §eOpens GUI\n" +
                    "§cSkySkipped §f:: §e/sm github §f- §eOpens official github page\n" +
                    "§cSkySkipped §f:: §e/sm crit §f- §eShows last critical hit\n" +
                    "§cSkySkipped §f:: §e/sm help §f- §eShow this message\n"
            );
        } else {
            UChat.chat("§cSkySkipped §f:: §eUsage:\n" +
                    "§cSkySkipped §f:: §e/sm §for§e /sm gui §f- §eOpens GUI\n" +
                    "§cSkySkipped §f:: §e/sm github §f- §eOpens official github page\n" +
                    "§cSkySkipped §f:: §e/sm crit §f- §eShows last critical hit\n" +
                    "§cSkySkipped §f:: §e/sm help §f- §eShows this message\n"
            );
        }
    }
}

File: src/main/kotlin/me/cephetir/skyskipped/commands/SkySkippedCommand.kt
/*
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * Version 2, December 2004
 *
 * Copyright (C) 2022 Cephetir
 *
 * Everyone is permitted to copy and distribute verbatim or modified
 * copies of this license document, and changing it is allowed as long
 * as the name is changed.
 *
 * DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
 * TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 */

package me.cephetir.skyskipped.commands

import gg.essential.api.EssentialAPI
import gg.essential.universal.UChat.chat
import gg.essential.universal.wrappers.message.UTextComponent
import me.cephetir.skyskipped.SkySkipped
import me.cephetir.skyskipped.config.Config
import me.cephetir.skyskipped.features.Features
import me.cephetir.skyskipped.features.impl.hacks.FailSafe
import me.cephetir.skyskipped.features.impl.hacks.PetMacro
import me.cephetir.skyskipped.gui.impl.GuiHudEditor
import me.cephetir.skyskipped.gui.impl.GuiItemSwap
import me.cephetir.skyskipped.utils.TextUtils.isNumeric
import net.minecraft.client.Minecraft
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.event.ClickEvent
import net.minecraft.event.HoverEvent
import net.minecraft.util.BlockPos
import net.minecraftforge.common.MinecraftForge

class SkySkippedCommand : CommandBase() {
    override fun getCommandName(): String {
        return "skyskipped"
    }

    override fun getCommandAliases(): List<String> {
        return listOf("sm")
    }

    override fun getCommandUsage(sender: ICommandSender): String {
        return "/$commandName help"
    }

    override fun getRequiredPermissionLevel(): Int {
        return 0
    }

    override fun addTabCompletionOptions(sender: ICommandSender, args: Array<String>, pos: BlockPos): List<String>? {
        return if (args.size == 1) getListOfStringsMatchingLastWord(args, "github", "crit", "help", "gui", "pet")
        else null
    }

    override fun processCommand(sender: ICommandSender, args: Array<String>) {
        if (args.isEmpty()) return EssentialAPI.getGuiUtil().openScreen(SkySkipped.config.gui())

        when (args[0]) {
            "gui" -> EssentialAPI.getGuiUtil().openScreen(SkySkipped.config.gui())
            "keybinds" -> EssentialAPI.getGuiUtil().openScreen(GuiItemSwap())
            "kb" -> EssentialAPI.getGuiUtil().openScreen(GuiItemSwap())
            "github" -> {
                val text = UTextComponent("§cSkySkipped §f:: §eGithub: §fhttps://github.com/Cephetir/SkySkipped")
                text.setHover(HoverEvent.Action.SHOW_TEXT, "§9Open URL in browser.")
                text.setClick(ClickEvent.Action.OPEN_URL, "https://github.com/Cephetir/SkySkipped")
                sender.addChatMessage(text)
            }
            "pet" -> {
                if ((args.size >= 2) && args[1].isNumeric() && args[1].toInt() > 0) {
                    val player = Minecraft.getMinecraft().thePlayer
                    if (Config.petsOverlay) Features.petsOverlay.auto = args[1].toInt()
                    else MinecraftForge.EVENT_BUS.register(PetMacro(args[1].toInt()))
                    player.sendChatMessage("/pets")
                } else chat("§cSkySkipped §f:: §4Specify pet index! Usage: /sm pet [pet index (start counting from 1)]")
            }
            "stop" -> {
                if (FailSafe.called4) {
                    FailSafe.called4 = false
                    chat("§cSkySkipped §f:: §eJacob failsafe stopped and won't reenable macro!")
                } else chat("§cSkySkipped §f:: §4Jacob failsafe wasn't triggered!")
            }
            "help" -> chat(
                """
                    §cSkySkipped §f:: §eUsage:
                    §cSkySkipped §f:: §e/sm §for§e /sm gui §f- §eOpens GUI
                    §cSkySkipped §f:: §e/sm keybinds §for§e /sm kb §f- §eOpens item swap keybinds GUI
                    §cSkySkipped §f:: §e/sm github §f- §eOpens official github page
                    §cSkySkipped §f:: §e/sm pet [pet index (start counting from 1)] §f- §eAuto select pet quickly
                    §cSkySkipped §f:: §e/sm trail [particle name] §f- §eSet trail particle
                    §cSkySkipped §f:: §e/sm hud §f- §eOpens hud editor GUI
                    §cSkySkipped §f:: §e/sm help §f- §eShow this message
                    """.trimIndent()
            )
            "dev" -> {
                SkySkipped.devMode = !SkySkipped.devMode
                chat("§cSkySkipped §f:: §eDev mode ${SkySkipped.devMode}")
            }
            "trail" -> {
                if (args.size >= 2) Config.trailParticle = args[1]
                else chat("§cSkySkipped §f:: §4Specify particle name!")
            }
            "hud" -> EssentialAPI.getGuiUtil().openScreen(GuiHudEditor())
            else -> chat(
                """
                    §cSkySkipped §f:: §eUsage:
                    §cSkySkipped §f:: §e/sm §for§e /sm gui §f- §eOpens GUI
                    §cSkySkipped §f:: §e/sm keybinds §for§e /sm kb §f- §eOpens item swap keybinds GUI
                    §cSkySkipped §f:: §e/sm github §f- §eOpens official github page
                    §cSkySkipped §f:: §e/sm pet [pet index (start counting from 1)] §f- §eAuto select pet very fast
                    §cSkySkipped §f:: §e/sm trail [particle name] §f- §eSet trail particle
                    §cSkySkipped §f:: §e/sm hud §f- §eOpens hud editor GUI
                    §cSkySkipped §f:: §e/sm help §f- §eShows this message
                    """.trimIndent()
            )
        }
    }
}